### DISCUSSION TITLE
support gatsby-plugin-image in @directus/gatsby-source-directus

### DISCUSSION THREAD

**Discussion by prKrumm** [directus/directus#6766] [Feature Requests]
Hi guys, I use Directus 8 with Gatsby and wanted to update Directus to Version 9. Unfortunately I am having trouble with images. I use the new plugin gatsby-plugin-image, which worked well with Directus 8.  e.g.:
``` 
 allDirectusLocation(filter: {city: {eq: $id}}) {
        nodes {
            background_image {
                localFile {
                    childImageSharp {
                        gatsbyImageData(layout: FULL_WIDTH, placeholder: TRACED_SVG)
                    }
                }
            }
        }
}
 ```
With Directus 9 it looks like:
```
directus {
    items {
      location {
        background_image {
          embed
          description
          charset
          id
          ...
        }
      }
    }
```

How can I use 
```
childImageSharp {
      gatsbyImageData(layout: FULL_WIDTH, placeholder: TRACED_SVG)
}
```

GatsbyImage is part of Gatsby's core, so I think should be part of the source plugin
Strapi (which I used before, but gladly made the move to Directus) has a source plugin in which they make use of the createRemoteFileNode .

**Comment by rijkvanzanten** [comment]
I personally have no clue (don't use Gatsby myself, nor was involved in the Gatsby source plugins)

@WoLfulus any thoughts? How does Gatsby treat the images?

**Reply by WoLfulus** [reply]
Someone already asked for file integration with Gatsby's filesystem, I'll take a look how it works

Right now it only returns what our GraphQL returns, no dynamic schema changes yet.

**Reply by YannickMol** [reply]
That was me on Discord! Looking forward to it 😃

**Comment by fatihky** [comment]
I use https://github.com/graysonhicks/gatsby-plugin-remote-images for fetching images from DatoCMS, it may be a good workaround. I didn't use directus's gatsby plugin so I don't have any idea about the main question but wanted to share this module as it seems your problem is similar to mine.

**Comment by cdanny358** [comment]
I was able to solve this using Gatsby's `gatsby-source-graphql` plugin and a snippet of code in the `gatsby-node.js` file. 

In `gatsby-config.js`, add this:

```
{
  resolve: 'gatsby-source-graphql',
  options: {
    url: '<your directus url>/graphql',
    typeName: 'Directus',
    fieldName: 'directus'
  }
}
```

In `gatsby-node.js`, add this:

```
const { createRemoteFileNode } = require('gatsby-source-filesystem')

exports.createResolvers = async({ actions, cache, createNodeId, createResolvers, store, reporter }) => {
  const { createNode } = actions

  await createResolvers({
    Directus_directus_files: {
      imageFile: {
        type: 'File',
        async resolve(source) {
          if (!source || !source.id) { return null }
          return await createRemoteFileNode({
            url: encodeURI(`<your directus url>/assets/${source.id}`),
            store,
            cache,
            createNode,
            createNodeId,
            reporter
          })
        }
      }
    }
  })
}
```

Now every image will have an `imageFile` field with the proper Gatsby fields, like this:

```
export const query = graphql`
  query {
    directus {
      my_collection {
        my_image {
          id
          imageFile {
            childImageSharp {
              gatsbyImageData
            }
          }
        }
      }
    }
  }
```

Note that the image should always request the id whether or not you need it, so that the resolver works properly.

**Comment by YannickMol** [comment]
@WoLfulus, just added a pull request (#5010) including @cdanny358's comments. Feel free to test it before merging!

**Reply by vpopat** [reply]
With this issue merged in v9.0.0-rc.59, does anything need to be added to gatsby-config.js or gatsby-node.js to get it to work ?

**Reply by benhaynes** [reply]
@WoLfulus ?

**Reply by WoLfulus** [reply]
changing your query should be enough

**Reply by vpopat** [reply]
My query is:
`{
  directus {
    product(filter: {slug: {_eq: "box"}}) {
      id
      name
      price
      slug
      featured {
        filename_disk
        imageFile {
          childImageSharp {
            gatsbyImageData
          }
        }
      }
    }
  }
}
`

`{
  "data": {
    "directus": {
      "product": [
        {
          "id": 1,
          "name": "Box",
          "price": 120,
          "slug": "box",
          "featured": {
            "filename_disk": "8dad5a23-ce8a-48f6-961b-5435d939e97c.jpeg",
            "imageFile": null
          }
        }
      ]
    }
  },
  "extensions": {}
}`

filaname_disk has a value but imageFile does not - any ideas ?

**Reply by cdanny358** [reply]
You probably need to request the `id` of `featured`, as noted in the comment above.

**Reply by vpopat** [reply]
Perfect! Thanks @cdanny358

**Comment by unmaskz** [comment]
Struggling to get this working - keeps giving issues:

`warn `createResolvers` passed resolvers for type `DirectusData_directus_files`
that doesn't exist in the schema. Use `createTypes` to add the type before
adding resolvers.`

` Field "image" must not have a selection since type "String" has no
subfields.`

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
